File: day09/puzzle.js
const fs = require("fs");
const readLines = (path) => {
  return fs
    .readFileSync(path, { encoding: 'utf8' })
    .split("\n")
    .filter(Boolean)
    .map((line) => line.split(" "));
}

const DIRECTIONS = {
  "L": [-1, 0], "R": [1, 0], "U": [0, 1], "D": [0, -1]
}

class Point {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  name() {
    return `${this.x},${this.y}`;
  }

  moveIn(direction) {
    const [dx, dy] = DIRECTIONS[direction];
    this.x += dx;
    this.y += dy;
  }

  keepAdjacencyWith(other) {
    const dx = other.x - this.x;
    const dy = other.y - this.y;
    if (Math.abs(dx) < 2 && Math.abs(dy) < 2) return;

    this.x += dx === 0 ? 0 : Math.floor(dx / Math.abs(dx));
    this.y += dy === 0 ? 0 : Math.floor(dy / Math.abs(dy));
  }
}

const puzzleInput = readLines(process.argv[2] || "input.txt");

const countUniqTailPositions = (knot) => {
  const visitSet = new Set();
  const head = knot[0];
  for (const [dir, steps] of puzzleInput) {
    for (let step = 0; step < Number(steps); step++) {
      head.moveIn(dir);
      knot.forEach((curr, i) => {
        if (i === 0) return;
        curr.keepAdjacencyWith(knot[i - 1])
      });
      visitSet.add(knot.at(-1).name());
    }
  }
  return visitSet.size
}

const shortKnot = [new Point(0, 0), new Point(0, 0)];
console.log(countUniqTailPositions(shortKnot)); // Part1

const longKnot = Array(10).fill(null).map(() => new Point(0, 0));
console.log(countUniqTailPositions(longKnot));  // Part2

File: day09/puzzle.rb
# frozen_string_literal: true

require "set"

@input = File.readlines(*ARGV, chomp: true).map(&:split)

Point = Struct.new(:x, :y) do
  def move_in(dir)
    case dir
    when "L" then self.x -= 1
    when "R" then self.x += 1
    when "D" then self.y -= 1
    when "U" then self.y += 1
    end
  end

  def keep_adjacency_with(other)
    dx = other.x - x
    dy = other.y - y
    return if dx.abs < 2 && dy.abs < 2

    self.x += dx.zero? ? 0 : dx / dx.abs
    self.y += dy.zero? ? 0 : dy / dy.abs
  end
end

def count_uniq_tail_positions(knot)
  visit_set = Set.new
  @input.each do |dir, count|
    count.to_i.times do
      head = knot[0]
      head.move_in(dir)
      knot.each_cons(2) { |prev, curr| curr.keep_adjacency_with(prev) }
      visit_set << knot[-1].clone
    end
  end
  visit_set.size
end

short_knot = Array.new(2) { Point.new(0, 0) }
p count_uniq_tail_positions(short_knot) # p1-> 5779

long_knot = Array.new(10) { Point.new(0, 0) }
p count_uniq_tail_positions(long_knot) # p2-> 2331
